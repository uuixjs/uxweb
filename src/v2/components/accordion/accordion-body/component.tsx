import { Background, BorderRadiusProps, Display, Layout } from "../../layout";
import { FC, ReactNode } from "react";

import { getAriaProps } from "lib";

export interface AccordionBodyProps {
  /**
   * @example Lorem ipsum dolor sit amet, consectetur adipiscing elit
   */
  children?: ReactNode;
  /**
   * @example true
   */
  isOpen?: boolean;
  borderRadius?: BorderRadiusProps;
}

/**
 * AccordionBody is a content region whose visibility should be
 * controlled by an AccordionHeader
 */
export const AccordionBody: FC<AccordionBodyProps> = (props) => (
  <Layout
    padding={2}
    background={Background.Base}
    elevation={1}
    display={props.isOpen ? Display.Block : Display.Hide}
    borderRadius={props.borderRadius}
    {...getAriaProps(props)}
  >
    {props.children}
  </Layout>
);

AccordionBody.displayName = "AccordionBody";
