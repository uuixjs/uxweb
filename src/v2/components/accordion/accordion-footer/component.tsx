import { Button, ButtonProps, ButtonType } from "../../button/button";
import { Display, JustifyContent, Layout } from "../../layout";

import { FC } from "react";
import { getAriaProps } from "lib/ui-utils";

export interface AccordionFooterProps {
  primaryButtonProps: ButtonProps;
  secondaryButtonProps?: ButtonProps;
}

/**
 * AccordionFooter should be placed within an AccordionBody and
 * composes either one or two call-to-action buttons with consistent placement.
 */
export const AccordionFooter: FC<AccordionFooterProps> = (props) => {
  const primaryButton = (
    // Order of attributes matters; later props will override earlier props
    <Button {...props.primaryButtonProps} variant={ButtonType.Primary} />
  );

  let secondaryButton;

  if (props.secondaryButtonProps) {
    secondaryButton = (
      // Order of attributes matters; later props will override earlier props
      <Button {...props.secondaryButtonProps} variant={ButtonType.Secondary} />
    );
  }

  return (
    <Layout
      display={Display.Flex}
      justifyContent={JustifyContent.End}
      margin={{ top: 2 }}
      {...getAriaProps(props)}
    >
      {secondaryButton && (
        <Layout margin={{ left: 1 }}>{secondaryButton}</Layout>
      )}

      <Layout margin={{ left: 1 }}>{primaryButton}</Layout>
    </Layout>
  );
};

AccordionFooter.displayName = "AccordionFooter";
