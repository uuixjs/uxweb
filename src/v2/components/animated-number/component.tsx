import { Component } from "react";
import CountUp from "react-countup";
import { styled } from "lib/ui-utils";

export interface AnimatedNumberProps {
  value: number;
  formatValue: (n: number) => string;
  initialValue?: number;
  duration?: number;
  children?: undefined;
  monospaced?: boolean;
}

export interface AnimatedNumberState {
  prevValue: number;
  currentValue: number;
}

export const ScAnimatedNumber = styled(CountUp)<{
  $monospaced: AnimatedNumberProps["monospaced"];
}>(({ $monospaced }) => {
  return (
    $monospaced && {
      "font-feature-settings": `"tnum"`,
    }
  );
});
export class AnimatedNumber extends Component<
  AnimatedNumberProps,
  AnimatedNumberState
> {
  public static defaultProps: Partial<AnimatedNumberProps> = {
    duration: 0.7,
    monospaced: true,
  };

  public constructor(props: AnimatedNumberProps) {
    super(props);
    this.state = {
      currentValue: props.value,
      prevValue:
        props.initialValue !== undefined ? props.initialValue : props.value,
    };
  }

  public componentDidUpdate() {
    if (this.props.value !== this.state.currentValue) {
      this.setState((s) => ({
        currentValue: this.props.value,
        prevValue: s.currentValue,
      }));
    }
  }

  public render() {
    const {
      duration,
      formatValue,
      initialValue,
      monospaced,
      value,
      ...props
    } = this.props;
    return (
      <ScAnimatedNumber
        $monospaced={monospaced}
        duration={duration}
        end={this.state.currentValue}
        start={this.state.prevValue}
        formattingFn={formatValue}
        {...props}
      />
    );
  }
}
