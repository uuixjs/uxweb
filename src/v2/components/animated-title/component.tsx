import { Component, Fragment } from "react";
import { keyframes, staticTokenRule, styled } from "lib/ui-utils";

const DELAY_INTERVAL_MS = 30;
const DELIMITER = " ";
const DISTANCE = 60;

export interface AnimatedTitleProps {
  children: string;
}

interface ScAnimatedTitleSegmentProps {
  index: number;
}

const ScAnimatedTitleEntrance = keyframes`
  0% {
    opacity: 0;
    transform: translateY(${DISTANCE + "px"});
  }
  0.1% {
    opacity: 1;
  }
  25% {
    transform: translateY(${DISTANCE + "px"});
  }
  43% {
    transform: translateY(${DISTANCE * 0.896 + "px"});
  }
  43.75% {
    transform: translateY(${DISTANCE * 0.567 + "px"});
  }
  44.5% {
    transform: translateY(${DISTANCE * 0.307 + "px"});
  }
  55% {
    transform: translateY(${DISTANCE * 0.045 + "px"});
  }
  100% {
    transform: translateY(${DISTANCE * 0 + "px"});
  }
`;

const ScAnimatedTitleSegment = styled.span<ScAnimatedTitleSegmentProps>`
  font-family: ${staticTokenRule("font-display")};
  display: inline-block;
  animation-name: ${ScAnimatedTitleEntrance};
  animation-duration: 385ms;
  animation-timing-function: linear;
  animation-iteration-count: 1;
  animation-fill-mode: both;
  animation-delay: ${({ index }) => DELAY_INTERVAL_MS * index + "ms"};
`;

export class AnimatedTitle extends Component<AnimatedTitleProps, {}> {
  public render() {
    const segments = this.props.children.split(DELIMITER);

    return segments.map((s, i) => (
      <Fragment key={s + i}>
        <ScAnimatedTitleSegment
          className="tw-animated-title__segment"
          index={i}
        >
          {s}
        </ScAnimatedTitleSegment>
        {i < segments.length && DELIMITER && <span>{DELIMITER}</span>}
      </Fragment>
    ));
  }
}
