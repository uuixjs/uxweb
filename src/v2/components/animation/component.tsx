import { AnimationEvent, FC, ReactNode } from "react";
import {
  bounce,
  bounceIn,
  bounceOut,
  fadeIn,
  fadeOut,
  slideInBottom,
  slideInLeft,
  slideInRight,
  slideInTop,
  slideOutBottom,
  slideOutLeft,
  slideOutRight,
  slideOutTop,
} from "./keyframes";
import { styleVariant, styled } from "lib/ui-utils";

export enum AnimationDelay {
  Short = "short",
  Medium = "medium",
  Long = "long",
  ExtraLong = "extra-long",
}

export enum AnimationDuration {
  Short = "short",
  Medium = "medium",
  Long = "long",
  ExtraLong = "extra-long",
}

export enum AnimationFillMode {
  Both = "both",
  None = "none",
  Forwards = "forwards",
  Backwards = "backwards",
}

export enum AnimationTiming {
  Ease = "ease",
  Linear = "linear",
  EaseIn = "ease-in",
  EaseOut = "ease-out",
  EaseInOut = "ease-in-out",
}

export enum AnimationType {
  Bounce = "bounce",
  BounceIn = "bounce-in",
  BounceOut = "bounce-out",
  FadeIn = "fade-in",
  FadeOut = "fade-out",
  SlideInTop = "slide-in-top",
  SlideInRight = "slide-in-right",
  SlideInBottom = "slide-in-bottom",
  SlideInLeft = "slide-in-left",
  SlideOutTop = "slide-out-top",
  SlideOutRight = "slide-out-right",
  SlideOutBottom = "slide-out-bottom",
  SlideOutLeft = "slide-out-left",
}

interface ScAnimationProps {
  $delay: AnimationProps["delay"];
  $duration: AnimationProps["duration"];
  $enabled: AnimationProps["enabled"];
  $fillMode: AnimationProps["fillMode"];
  $loop: AnimationProps["loop"];
  $timing: AnimationProps["timing"];
  $type: AnimationProps["type"];
}

const ScAnimation = styled.div<ScAnimationProps>`
  animation-name: ${(props) =>
    props.$enabled &&
    styleVariant("$type", {
      [AnimationType.Bounce]: bounce,
      [AnimationType.BounceIn]: bounceIn,
      [AnimationType.BounceOut]: bounceOut,
      [AnimationType.FadeIn]: fadeIn,
      [AnimationType.FadeOut]: fadeOut,
      [AnimationType.SlideInBottom]: slideInBottom,
      [AnimationType.SlideInLeft]: slideInLeft,
      [AnimationType.SlideInRight]: slideInRight,
      [AnimationType.SlideInTop]: slideInTop,
      [AnimationType.SlideOutBottom]: slideOutBottom,
      [AnimationType.SlideOutLeft]: slideOutLeft,
      [AnimationType.SlideOutRight]: slideOutRight,
      [AnimationType.SlideOutTop]: slideOutTop,
    })};

  animation-delay: ${styleVariant("$delay", {
    [AnimationDelay.Short]: "0.15s",
    [AnimationDelay.Medium]: "0.3s",
    [AnimationDelay.Long]: "0.6s",
    [AnimationDelay.ExtraLong]: "1s",
  })};

  animation-duration: ${styleVariant("$duration", {
    [AnimationDuration.Short]: "0.15s",
    [AnimationDuration.Medium]: "0.3s",
    [AnimationDuration.Long]: "0.6s",
    [AnimationDuration.ExtraLong]: "1s",
  })};

  animation-fill-mode: ${styleVariant("$fillMode", {
    [AnimationFillMode.None]: "none",
    [AnimationFillMode.Forwards]: "forwards",
    [AnimationFillMode.Backwards]: "backwards",
    [AnimationFillMode.Both]: "both",
  })};

  animation-timing-function: ${styleVariant("$timing", {
    [AnimationTiming.Ease]: "ease",
    [AnimationTiming.EaseIn]: "ease-in",
    [AnimationTiming.EaseInOut]: "ease-in-out",
    [AnimationTiming.EaseOut]: "ease-out",
    [AnimationTiming.Linear]: "linear",
  })};

  animation-iteration-count: ${(props) => props.$loop && "infinite"};
`;

export interface AnimationProps {
  /**
   * The content to render inside the animation.
   *
   * @example <CoreText fontSize={FontSize.Size4}>Toggle the "enabled" property to animate.</CoreText>
   */
  children?: ReactNode;
  /** The delay before the animation is triggered. */
  delay?: AnimationDelay | "short" | "medium" | "long" | "extra-long"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** The duration of the animation. */
  duration?: AnimationDuration | "short" | "medium" | "long" | "extra-long"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /**
   * Used to trigger the animation. Toggling this prop will fire the animation.
   *
   * @example true
   */
  enabled?: boolean;
  /** Applies the CSS `animation-fill-mode` property. */
  fillMode?: AnimationFillMode | "both" | "none" | "forwards" | "backwards"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** When enabled, loop the animation. */
  loop?: boolean;
  /** Event called when the animation start. */
  onAnimationStart?: (e: AnimationEvent<HTMLDivElement>) => void;
  /** Event called when the animation ends. */
  onAnimationEnd?: (e: AnimationEvent<HTMLDivElement>) => void;
  /** The easing function that will be used for the animation. */
  timing?:
    | AnimationTiming
    | "ease"
    | "linear"
    | "ease-in"
    | "ease-out"
    | "ease-in-out"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** The type of animation that will occur. */
  type:
    | AnimationType
    | "bounce"
    | "bounce-in"
    | "bounce-out"
    | "fade-in"
    | "fade-out"
    | "slide-in-top"
    | "slide-in-right"
    | "slide-in-bottom"
    | "slide-in-left"
    | "slide-out-top"
    | "slide-out-right"
    | "slide-out-bottom"
    | "slide-out-left"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

/**
 * Animations provide context and cues to the user interface.
 *
 * @deprecated Deprecated in favor of the new <Transition> component which was introduced in 5.5.0.
 * Please migrate your usage of <Animation> to the new <Transition> component. If you have a use case
 * which is not met by the new component, please reach out on Slack.
 * [See New Component](https://design.xarth.tv/user-interface/patterns/transition)
 * [Release Notes](https://git.xarth.tv/core-ui/core-ui/releases/tag/v5.5.0)
 */
export const Animation: FC<AnimationProps> = ({
  delay,
  duration = AnimationDuration.Short,
  enabled,
  fillMode = AnimationFillMode.Both,
  loop,
  timing = AnimationTiming.Ease,
  type,
  ...props
}) => {
  return (
    <ScAnimation
      $delay={delay}
      $duration={duration}
      $enabled={enabled}
      $fillMode={fillMode}
      $loop={loop}
      $timing={timing}
      $type={type}
      className="tw-animation"
      data-a-target="tw-animation-target"
      {...props}
    />
  );
};

Animation.displayName = "Animation";
