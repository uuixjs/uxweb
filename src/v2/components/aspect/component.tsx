import { FC, ReactNode } from "react";

import { styled } from "lib/ui-utils";

export enum AspectRatio {
  Aspect21x9 = "tw-aspect--21x9",
  Aspect16x9 = "tw-aspect--16x9",
  Aspect4x3 = "tw-aspect--4x3",
  Aspect1x1 = "tw-aspect--1x1",
  Aspect3x4 = "tw-aspect--3x4",
  Aspect3x2 = "tw-aspect--3x2",
  // Leaving BoxArt as a semantic alias for now.
  BoxArt = "tw-aspect--3x4",
}

const AspectRatioValues = {
  [AspectRatio.Aspect21x9]: 21 / 9,
  [AspectRatio.Aspect16x9]: 16 / 9,
  [AspectRatio.Aspect4x3]: 4 / 3,
  [AspectRatio.Aspect1x1]: 1,
  [AspectRatio.Aspect3x4]: 3 / 4,
  [AspectRatio.Aspect3x2]: 3 / 2,
  [AspectRatio.BoxArt]: 3 / 4,
};

export enum Align {
  Bottom = 1,
  Center = 2,
  Top = 3,
}

interface ScAspectRatioProps {
  /**
   * Align the content within the component (if the content has a different
   * aspect ratio than the component).
   */
  align?: Align;
  /**
   * Show overflowing content (if the content has a different aspect ratio
   * than the component).
   */
  allowOverflow?: boolean;
}

function getAlign({ align }: ScAspectRatioProps): string {
  switch (align) {
    case Align.Bottom:
      return "bottom: 0;";
    case Align.Center:
      return "top: 50%; transform: translateY(-50%)";
    default:
      return "top: 0;";
  }
}

// @ts-ignore
export const ScAspectSpacer = styled.div<{ ratio: number | any }>`
  ${(props) => `padding-bottom: ${(1 / Math.abs(props.ratio)) * 100}%`}
`;

export const ScAspectRatio = styled.div<ScAspectRatioProps>`
  position: relative;
  width: 100%;
  overflow: ${(props) => (props.allowOverflow ? "visible" : "hidden")};

  /* stylelint-disable selector-max-universal, selector-max-type */
  & > *:not(${ScAspectSpacer}) {
    position: absolute;
    left: 0;
    width: 100%;
    min-height: 100%;
    ${getAlign}
  }
`;

export interface AspectProps extends ScAspectRatioProps {
  /**
   * The desired aspect ratio. If expressed as a number this value is width divided by height.
   * For example, a 3x4 ratio (box art) is calculated as 3 divided by 4 which equals 0.75
   *
   * @example 0.75
   */
  ratio?: AspectRatio | number;
  /**
   * The content to be contained within the aspect.
   *
   * @example <img src="https://static-cdn.jtvnw.net/s3_vods/twitchpresents/159083707/cf22b5ad-ce79-454b-bd91-c9831579ecfc/thumb/index-0000000000-1280x720.jpg" />
   */
  children?: ReactNode;
  /**
   * Deprecated alias for allowOverflow.
   *
   * @deprecated
   */
  overflow?: boolean;
}

/**
 * Aspect is used to set the height of an element based on its width for
 * applications such as box art and thumbnail images.
 */
export const Aspect: FC<AspectProps> = ({
  ratio = AspectRatioValues[AspectRatio.Aspect16x9],
  children,
  allowOverflow,
  overflow,
  ...aspectProps
}) => {
  return (
    <ScAspectRatio
      className="tw-aspect"
      allowOverflow={allowOverflow ?? overflow}
      {...aspectProps}
    >
      <ScAspectSpacer
        ratio={
          ratio && typeof ratio === "string" ? AspectRatioValues[ratio] : ratio
        }
      />
      {children}
    </ScAspectRatio>
  );
};

Aspect.defaultProps = {
  align: Align.Top,
  // This uses the deprecated prop to enable proper fallback from the new
  // `allowOverflow`. Once `overflow` is fully removed, this should switch to
  // `allowOverflow`.
  overflow: false,
};

Aspect.displayName = "Aspect";
