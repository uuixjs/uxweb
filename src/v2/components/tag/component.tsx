import {
  CoreInteractive,
  CoreInteractiveElement,
  CoreInteractivePublicProps,
} from "../core-interactive";
import { ForwardRefRenderFunction, forwardRef } from "react";
import { SVG, SVGAsset } from "../svg";
import {
  css,
  focusVisible,
  hoverCss,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "lib/ui-utils";

import { withOverlayContext } from "../overlay-region/context";

interface ScTagProps {
  $overlay: TagProps["overlay"];
}

const overlayStyles = css`
  color: ${themeTokenRule("color-text-overlay")};
  background-color: ${staticTokenRule("color-opac-b-6")};
  border-color: ${themeTokenRule("color-border-input-overlay")};

  &:not(:disabled) {
    ${focusVisible`
      border-color: ${themeTokenRule("color-border-input-overlay-focus")};
    `}

    ${hoverCss`
      text-decoration: none;
      color: ${themeTokenRule("color-text-overlay")};
      background-color: ${themeTokenRule(
        "color-background-interactable-overlay-hover",
      )};
    `}

    /* Active must be below Hover to take precedence */
    &:active {
      background-color: ${themeTokenRule(
        "color-background-interactable-overlay-active",
      )};
    }
  }
`;

const ScTag = styled(CoreInteractive)<ScTagProps>`
  display: inline-flex;
  border-radius: ${staticTokenRule("border-radius-rounded")};
  font-weight: ${staticTokenRule("font-weight-semibold")};
  color: ${themeTokenRule("color-text-tag")};
  background-color: ${themeTokenRule("color-background-tag-default")};
  border: ${staticTokenRule("border-width-tag")} solid transparent;
  height: 2rem;

  &:not(:disabled) {
    ${focusVisible`
      border-style: solid;
      border-width:${staticTokenRule("border-width-tag")};
      border-color: ${themeTokenRule("color-border-input-overlay-focus")};
      background-color: ${themeTokenRule("color-background-tag-hover")};
    `}

    ${hoverCss`
      text-decoration: none;
      color: ${themeTokenRule("color-text-tag")};
      background-color: ${themeTokenRule("color-background-tag-hover")};
    `}

    /* Active must be below Hover to take precedence */
    &:active {
      background-color: ${themeTokenRule("color-background-tag-active")};
    }
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  ${({ $overlay }) => $overlay && overlayStyles}
`;

const ScTagContent = styled.div<{ action: TagProps["action"] }>`
  display: flex;
  align-items: center;
  font-size: ${staticTokenRule("font-size-7")};
  padding: 0 0.8rem;
  padding-right: ${({ action }) => (action ? "0.4rem" : undefined)};
`;

const ScTagIconWrapper = styled.div`
  display: flex;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
`;

export enum TagAction {
  Remove = "remove",
  Add = "add",
}

export interface TagProps extends CoreInteractivePublicProps {
  action?: TagAction | "remove" | "add"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
  /**
   * Text to render within the tag.
   *
   * @example Competitive
   */
  label: string;
  overlay?: boolean;
}

export const TagComponent: ForwardRefRenderFunction<
  CoreInteractiveElement,
  TagProps
> = ({ label, action, overlay, ...tagProps }, ref) => {
  return (
    <ScTag className="tw-tag" $overlay={overlay} {...tagProps} ref={ref}>
      <ScTagContent action={action}>
        {label}
        {action && (
          <ScTagIconWrapper>
            <SVG
              fill
              asset={
                action === TagAction.Add
                  ? SVGAsset.Smallplus
                  : SVGAsset.Smallhide
              }
            />
          </ScTagIconWrapper>
        )}
      </ScTagContent>
    </ScTag>
  );
};

TagComponent.displayName = "Tag";
const ComponentWithRef = forwardRef(TagComponent);
const ComponentWithOverlayContext = withOverlayContext<TagProps, HTMLElement>(
  ComponentWithRef,
);
export { ComponentWithOverlayContext as Tag };
