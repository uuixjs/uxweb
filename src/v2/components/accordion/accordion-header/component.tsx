import { Align, Aspect, AspectRatio } from "../../aspect";
import {
  AlignItems,
  Background,
  BorderRadiusProps,
  Color,
  Display,
  FlexDirection,
  InjectLayout,
  JustifyContent,
  Layout,
  Overflow,
  Position,
  ZIndex,
} from "../../layout";
import { Component, MouseEvent } from "react";
import { CoreImage, CoreImageProps } from "../../core-image";
import { CoreText, TextType } from "../../core-text";
import {
  ExpandableIndicator,
  ExpandableIndicatorDirection,
} from "../../expandable-indicator";
import { PaddingValue, getAriaProps, styled } from "lib";
import { Title, TitleSize } from "../../title";

import { Interactable } from "../../interactable";

export enum AccordionHeaderSize {
  Default = "default",
  Narrow = "narrow",
}

export interface AccordionDefaultHeaderProps {
  index?: number;
  /**
   * @example What is thy bidding?
   */
  title: string;
  /**
   * @example Reach 10 followers
   */
  description?: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  imageProps?: CoreImageProps;
  imageRatio?: AspectRatio.Aspect1x1 | AspectRatio.Aspect3x4;
  backgroundImageProps?: CoreImageProps;
  imageLabelOverlay?: string;
  label?: string;
  borderRadius?: BorderRadiusProps;
  size?: AccordionHeaderSize.Default;
}

export interface AccordionNarrowHeaderProps {
  index?: number;
  title: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  borderRadius?: BorderRadiusProps;
  size: AccordionHeaderSize.Narrow;
}

const defaultHeight = "8rem";
const narrowHeight = "5rem";

const ScAccordionHeaderContents = styled(Layout)<{
  $size: AccordionHeaderSize;
}>(({ $size }) =>
  $size && $size === AccordionHeaderSize.Default
    ? `min-height: ${defaultHeight};`
    : `height: ${narrowHeight};`,
);

const ScAccordionHeaderImageContainer = styled(Layout)<{
  $ratio: AccordionDefaultHeaderProps["imageRatio"];
}>(({ $ratio }) =>
  $ratio && $ratio === AspectRatio.Aspect3x4
    ? `min-width: calc(${defaultHeight} * (3/4));`
    : `min-width: ${defaultHeight};`,
);

const ScAccordionHeaderBackgroundImage = styled(CoreImage)`
  opacity: 0.5;
  object-fit: cover;
  object-position: center;
`;

export type AccordionHeaderProps =
  | AccordionDefaultHeaderProps
  | AccordionNarrowHeaderProps;

/**
 * An AccordionHeader is an interactable region which should be used to toggle
 * the visibility of additional information.
 */
export class AccordionHeader extends Component<AccordionHeaderProps, {}> {
  public render() {
    return (
      <Layout
        className="accordion-header"
        background={Background.Base}
        elevation={1}
        position={Position.Relative}
        borderRadius={this.props.borderRadius}
        {...getAriaProps(this.props)}
      >
        {this.renderBackgroundImage()}

        <Interactable
          onClick={this.onClick}
          aria-expanded={this.props.isOpen}
          borderRadius={this.props.borderRadius}
        >
          <ScAccordionHeaderContents
            $size={
              this.props.size === AccordionHeaderSize.Default ||
              this.props.size === undefined
                ? AccordionHeaderSize.Default
                : AccordionHeaderSize.Narrow
            }
            display={Display.Flex}
            position={Position.Relative}
            borderRadius={this.props.borderRadius}
            overflow={Overflow.Hidden}
          >
            {this.renderImageAndLabel()}
            {this.renderTitles()}
            {this.renderLabel()}
            {this.renderCaret()}
          </ScAccordionHeaderContents>
        </Interactable>
      </Layout>
    );
  }

  private getYPadding = (): PaddingValue => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return 2;
    }
    return this.props.description ? 2 : 3;
  };

  private renderCaret = () => {
    return (
      <Layout
        flexGrow={0}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.Center}
        padding={{ y: this.getYPadding(), right: 2 }}
      >
        <ExpandableIndicator
          open={this.props.isOpen}
          openDirection={ExpandableIndicatorDirection.Up}
          closedDirection={ExpandableIndicatorDirection.Down}
        />
      </Layout>
    );
  };

  private renderTitles = () => {
    return (
      <Layout
        flexGrow={1}
        flexShrink={1}
        display={Display.Flex}
        flexDirection={FlexDirection.Column}
        alignItems={AlignItems.Start}
        justifyContent={JustifyContent.Center}
        overflow={Overflow.Hidden}
        padding={{ y: this.getYPadding(), left: 2, right: 1 }}
      >
        <Layout fullWidth>
          <Title type={TextType.H3} size={TitleSize.Small} ellipsis>
            {this.props.title}
          </Title>
          {!isAccordionHeaderNarrowProps(this.props) && this.props.description && (
            <CoreText color={Color.Alt2} ellipsis>
              {this.props.description}
            </CoreText>
          )}
        </Layout>
      </Layout>
    );
  };

  private renderBackgroundImage = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.backgroundImageProps) {
      return;
    }

    return (
      <InjectLayout
        className="accordion-header__background-image"
        position={Position.Absolute}
        attachTop
        attachLeft
        fullWidth
        fullHeight
        borderRadius={this.props.borderRadius}
      >
        <ScAccordionHeaderBackgroundImage
          {...this.props.backgroundImageProps}
        />
      </InjectLayout>
    );
  };

  private renderLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.label) {
      return;
    }

    return (
      <Layout
        padding={{ right: 2 }}
        flexGrow={1}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.End}
      >
        <CoreText>{this.props.label}</CoreText>
      </Layout>
    );
  };

  private renderImageAndLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.imageProps && !this.props.imageLabelOverlay) {
      return;
    }

    return (
      <ScAccordionHeaderImageContainer
        $ratio={this.props.imageRatio}
        flexGrow={0}
        flexShrink={0}
        background={Background.Alt2}
        position={Position.Relative}
      >
        {this.props.imageLabelOverlay && (
          <Layout
            position={Position.Relative}
            fullHeight
            zIndex={ZIndex.Above}
            display={Display.Flex}
            alignItems={AlignItems.Center}
            justifyContent={JustifyContent.Center}
            padding={1}
            color={this.props.imageProps ? Color.Overlay : Color.Inherit}
            background={this.props.imageProps ? Background.Overlay : undefined}
          >
            <Title>{this.props.imageLabelOverlay}</Title>
          </Layout>
        )}
        {this.props.imageProps && (
          <Layout
            position={Position.Absolute}
            fullWidth
            fullHeight
            attachTop
            attachLeft
            overflow={Overflow.Hidden}
          >
            <Aspect
              ratio={this.props.imageRatio || AspectRatio.Aspect1x1}
              align={Align.Center}
            >
              <CoreImage {...this.props.imageProps} />
            </Aspect>
          </Layout>
        )}
      </ScAccordionHeaderImageContainer>
    );
  };

  private onClick = (event: MouseEvent<HTMLElement>) => {
    if (this.props.onClick !== undefined) {
      this.props.onClick(event, this.props.index);
    }
  };
}

function isAccordionHeaderNarrowProps(
  props: AccordionDefaultHeaderProps | AccordionNarrowHeaderProps,
): props is AccordionNarrowHeaderProps {
  return props.size === AccordionHeaderSize.Narrow;
}
