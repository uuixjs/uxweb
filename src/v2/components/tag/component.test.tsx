import { shallow } from "enzyme";
import { setupShallowTest } from "../../tests/helpers";
import { Tag, TagAction, TagComponent, TagProps } from "./component";

function getRequiredProps(): TagProps {
  return {
    label: "Test label",
  };
}

function getOptionalProps(): TagProps {
  return {
    disabled: true,
    label: "Test label",
    onClick: jest.fn(),
    tabIndex: 0,
    autoFocus: true,
    download: "download",
    linkTo: "https://example.com",
    targetBlank: true,
    title: "test title",
  };
}

const setupShallow = setupShallowTest(TagComponent, getRequiredProps);

describe("Tag", () => {
  it("renders tag component with only required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders tag component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("renders tag component with an actionType prop", () => {
    const { wrapper } = setupShallow({ action: TagAction.Add });
    expect(wrapper).toMatchSnapshot();
  });

  it("passes the aria label to the html", () => {
    const wrapper = shallow(
      <Tag aria-label="this is my aria label" {...getRequiredProps()} />,
    );
    expect(
      wrapper.html().match(/aria-label="this is my aria label"/)?.length,
    ).toBe(1);
  });
});
