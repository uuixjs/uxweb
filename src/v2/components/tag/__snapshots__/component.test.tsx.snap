// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders tag component with all props 1`] = `
<Styled(Component)
  autoFocus={true}
  className="tw-tag"
  disabled={true}
  download="download"
  linkTo="https://example.com"
  onClick={[MockFunction]}
  tabIndex={0}
  targetBlank={true}
  title="test title"
>
  <styled.div>
    Test label
  </styled.div>
</Styled(Component)>
`;

exports[`Tag renders tag component with an actionType prop 1`] = `
<Styled(Component)
  className="tw-tag"
>
  <styled.div
    action="add"
  >
    Test label
    <styled.div>
      <SVG
        asset="Smallplus"
        fill={true}
      />
    </styled.div>
  </styled.div>
</Styled(Component)>
`;

exports[`Tag renders tag component with only required props 1`] = `
<Styled(Component)
  className="tw-tag"
>
  <styled.div>
    Test label
  </styled.div>
</Styled(Component)>
`;
